src/drivers/TwiMaster.cpp

```diff
@@ -61,9 +62,11 @@ void TwiMaster::Init() {
 }
 
 TwiMaster::ErrorCodes TwiMaster::Read(uint8_t deviceAddress, uint8_t registerAddress, uint8_t *data, size_t size) {
-  xSemaphoreTake(mutex, portMAX_DELAY);
+  // this is causing an error when came from sleep
```

Comment:
I think we should analyze this is more details : the bus can be accessed by multiple tasks (Display for the touch and HeartRate for the heart rate sensor). Without this mutex, 2 tasks can stsart Read operations on the bus at the same time, which will cause undefined behaviors.

There's probably an issue is the Sleep/Wakeup order: 
 - the bus was still in use when going to sleep (a transaction couldn't finish because it went to sleep)
 - a task started a transaction before the bus was woken up

Reply:
yes that point was difficult, in my fork is working fine, but here I couldn’t get him to read the touch ... I didn’t investigate much, if anyone can find out why,

Reply:
I'll definitely investigate this :)

Reply:
I think the issue comes from `SystemTask::OnTouchEvent()` (SystemTask.cpp:271). This method is called in the context of the IRQ from the touch panel. It's not called in the context of the scheduler of FreeRTOS. 

When the wake up on double-tap is enabled, it receives the interrupt from the touchpanel and checks that the touch event is a double tap by reading the status from the touch panel. The method TwiMaster::Read() tries to get the semaphore, but freertos throws an assert because the scheduler is suspended (sleep mode). It probably does so to prevent deadlocks in case the semaphore were not available at that time.

I think about 2 solutions to work around this issue:
- Add *_FromIRQ() methods in CST816S and TwiMaster and remove the semaphore from these implementations
- Instead of checking the touch event in `SystemTask::OnTouchEvent()`, send a message to the SystemTask so that SystemTask checks the event and decides to wake up or stay asleep. In this case, SystemTask would run in the context of the scheduler and everything would be fine.

I think 2nd solution is the best, even if it's a bit more complex to implement.

Reply:
Nice, the second solution worked.